### PR TITLE
Fixes #29999 - Manage cvv export history

### PR DESCRIPTION
**Pull request by parthaa**
Fixes #29999 - Manage cvv export history

This commit adds new functionality to the pulp3 export api. Changes
include

* Create the ‘ContentViewExportHistory’ entity with
** id, importing_server, content-view version, path

This entity will used for bookkeeping for incrementals
** Also includes API to get a list of export histories

* Add a hammer content-view version export-history

Includes controller bindings in addition

### CODE REVIEW

**Comment by parthaa** [comment]
Would recommend looking at it from the most recent commit (since this PR contains code that'll get merged after pulpcore 3.6)

**Comment by parthaa** [comment]
Depends on https://github.com/Katello/katello/pull/8850

**Comment by parthaa** [comment]
Export with the following command
Note the new `destination-server` parameter

```
$ hammer  content-view version export-legacy --id=2 --destination-server=brew.example.com
[.....................................................................................................................................................................................................................................] [100%]
```
An export history entry should be created.
You should be able to run the following curl commands.

```
$ curl -X GET -s -k -u admin:changeme "http://$(hostname -f):3000/katello/api/content_view_versions/export_histories"|jq
{
  "total": 3,
  "subtotal": 3,
  "page": 1,
  "per_page": 20,
  "error": null,
  "search": null,
  "sort": {
    "by": "id",
    "order": "asc"
  },
  "results": [
    {
      "destination_server": "cool",
      "path": "/var/lib/pulp/exports/Default_Organization/cv/1.0/cool/1",
      "id": 1,
      "created_at": "2020-08-19 20:17:20 UTC",
      "updated_at": "2020-08-19 20:17:20 UTC",
      "content_view_version": "cv 1.0",
      "content_view_version_id": 2
    },
    {
      "destination_server": "downstream.redhat.com",
      "path": "/var/lib/pulp/exports/Default_Organization/cv/1.0/downstream.redhat.com/1",
      "id": 2,
      "created_at": "2020-08-19 22:06:08 UTC",
      "updated_at": "2020-08-19 22:06:08 UTC",
      "content_view_version": "cv 1.0",
      "content_view_version_id": 2
    },
    {
      "destination_server": "brew.example.com",
      "path": "/var/lib/pulp/exports/Default_Organization/cv/1.0/brew.example.com/1",
      "id": 3,
      "created_at": "2020-08-19 22:22:33 UTC",
      "updated_at": "2020-08-19 22:22:33 UTC",
      "content_view_version": "cv 1.0",
      "content_view_version_id": 2
    }
  ]
}

#with org id
$ curl -X GET -s -k -u admin:changeme "http://$(hostname -f):3000/katello/api/content_view_versions/export_histories?organization_id=1"|jq
{
  "total": 3,
  "subtotal": 3,
  "page": 1,
  "per_page": 20,
  "error": null,
  "search": null,
  "sort": {
    "by": "id",
    "order": "asc"
  },
  "results": [
    {
      "destination_server": "cool",
      "path": "/var/lib/pulp/exports/Default_Organization/cv/1.0/cool/1",
      "id": 1,
      "created_at": "2020-08-19 20:17:20 UTC",
      "updated_at": "2020-08-19 20:17:20 UTC",
      "content_view_version": "cv 1.0",
      "content_view_version_id": 2
    },
    {
      "destination_server": "downstream.redhat.com",
      "path": "/var/lib/pulp/exports/Default_Organization/cv/1.0/downstream.redhat.com/1",
      "id": 2,
      "created_at": "2020-08-19 22:06:08 UTC",
      "updated_at": "2020-08-19 22:06:08 UTC",
      "content_view_version": "cv 1.0",
      "content_view_version_id": 2
    },
    {
      "destination_server": "brew.example.com",
      "path": "/var/lib/pulp/exports/Default_Organization/cv/1.0/brew.example.com/1",
      "id": 3,
      "created_at": "2020-08-19 22:22:33 UTC",
      "updated_at": "2020-08-19 22:22:33 UTC",
      "content_view_version": "cv 1.0",
      "content_view_version_id": 2
    }
  ]
}

# search by destination server 
$ curl -X GET -s -k -u admin:changeme "http://$(hostname -f):3000/katello/api/content_view_versions/export_histories?destination_server=cool"|jq
{
  "total": 1,
  "subtotal": 1,
  "page": 1,
  "per_page": 20,
  "error": null,
  "search": null,
  "sort": {
    "by": "id",
    "order": "asc"
  },
  "results": [
    {
      "destination_server": "cool",
      "path": "/var/lib/pulp/exports/Default_Organization/cv/1.0/cool/1",
      "id": 1,
      "created_at": "2020-08-19 20:17:20 UTC",
      "updated_at": "2020-08-19 20:17:20 UTC",
      "content_view_version": "cv 1.0",
      "content_view_version_id": 2
    }
  ]
}

# search by destination server and organization id
$ curl -X GET -s -k -u admin:changeme "http://$(hostname -f):3000/katello/api/content_view_versions/export_histories?destination_server=brew.example.com&organization_id=1"|jq
{
  "total": 1,
  "subtotal": 1,
  "page": 1,
  "per_page": 20,
  "error": null,
  "search": null,
  "sort": {
    "by": "id",
    "order": "asc"
  },
  "results": [
    {
      "destination_server": "brew.example.com",
      "path": "/var/lib/pulp/exports/Default_Organization/cv/1.0/brew.example.com/1",
      "id": 3,
      "created_at": "2020-08-19 22:22:33 UTC",
      "updated_at": "2020-08-19 22:22:33 UTC",
      "content_view_version": "cv 1.0",
      "content_view_version_id": 2
    }
  ]
}


# By CVV Id
$ curl -X GET -s -k -u admin:changeme "http://$(hostname -f):3000/katello/api/content_view_versions/export_histories?content_view_version_id=2"|jq
{
  "total": 3,
  "subtotal": 3,
  "page": 1,
  "per_page": 20,
  "error": null,
  "search": null,
  "sort": {
    "by": "id",
    "order": "asc"
  },
  "results": [
    {
      "destination_server": "cool",
      "path": "/var/lib/pulp/exports/Default_Organization/cv/1.0/cool/1",
      "id": 1,
      "created_at": "2020-08-19 20:17:20 UTC",
      "updated_at": "2020-08-19 20:17:20 UTC",
      "content_view_version": "cv 1.0",
      "content_view_version_id": 2
    },
    {
      "destination_server": "downstream.redhat.com",
      "path": "/var/lib/pulp/exports/Default_Organization/cv/1.0/downstream.redhat.com/1",
      "id": 2,
      "created_at": "2020-08-19 22:06:08 UTC",
      "updated_at": "2020-08-19 22:06:08 UTC",
      "content_view_version": "cv 1.0",
      "content_view_version_id": 2
    },
    {
      "destination_server": "brew.example.com",
      "path": "/var/lib/pulp/exports/Default_Organization/cv/1.0/brew.example.com/1",
      "id": 3,
      "created_at": "2020-08-19 22:22:33 UTC",
      "updated_at": "2020-08-19 22:22:33 UTC",
      "content_view_version": "cv 1.0",
      "content_view_version_id": 2
    }
  ]
}

By Invalid cvv id
$ curl -X GET -s -k -u admin:changeme "http://$(hostname -f):3000/katello/api/content_view_versions/export_histories?content_view_version_id=3"|jq
{
  "total": 0,
  "subtotal": 0,
  "page": 1,
  "per_page": 20,
  "error": null,
  "search": null,
  "sort": {
    "by": "id",
    "order": "asc"
  },
  "results": []
}

```

File: app/controllers/katello/api/v2/content_view_versions_controller.rb [hunk 1]
```diff
@@ -82,8 +82,22 @@ def republish_repositories
       respond_for_async :resource => task
     end
 
+    api :GET, "/content_view_versions/export_histories", N_("Show the export history for a content view version")
+    param :content_view_version_id, :number, :desc => N_("Content view version identifier"), :required => false
+    param :destination_server, String, :desc => N_("Destination Server name")
+    param :organization_id, :number, :desc => N_("Organization identifier")
+    def export_histories
+      history = ContentViewVersionExportHistory.readable
+      history = history.where(:content_view_version_id => params[:content_view_version_id]) unless params[:content_view_version_id].blank?
```
Comment (jlsherrill) [inline app/controllers/katello/api/v2/content_view_versions_controller.rb:91]: would filtering by content view be more useful than by version?  or at least having it as an option too?

File: app/controllers/katello/api/v2/content_view_versions_controller.rb [hunk 2]
```diff
@@ -82,8 +82,22 @@ def republish_repositories
       respond_for_async :resource => task
     end
 
+    api :GET, "/content_view_versions/export_histories", N_("Show the export history for a content view version")
+    param :content_view_version_id, :number, :desc => N_("Content view version identifier"), :required => false
+    param :destination_server, String, :desc => N_("Destination Server name")
+    param :organization_id, :number, :desc => N_("Organization identifier")
```
Comment (jlsherrill) [inline app/controllers/katello/api/v2/content_view_versions_controller.rb:88]: these two prams seem not-required either

File: app/lib/actions/pulp3/content_view_version/create_exporter.rb [hunk 3]
```diff
@@ -5,12 +5,14 @@ class CreateExporter < Pulp3::Abstract
...
           param :content_view_version_id, Integer
+          param :destination_server, String
         end
 
         def run
           cvv = ::Katello::ContentViewVersion.find(input[:content_view_version_id])
           output[:exporter_data] = ::Katello::Pulp3::ContentViewVersion.new(smart_proxy: smart_proxy,
-                                                                            content_view_version: cvv).create_exporter
+                                                                            content_view_version: cvv,
+                                                                            destination_server: input[:destination_server]).create_exporter
```
Comment (jlsherrill) [inline app/lib/actions/pulp3/content_view_version/create_exporter.rb:15]: destination_server is an attribute on content_view_version?

Comment (jlsherrill) [reply app/lib/actions/pulp3/content_view_version/create_exporter.rb:15]: ohh thats the content view version service class...  Thats confusing ;)

Comment (jlsherrill) [reply app/lib/actions/pulp3/content_view_version/create_exporter.rb:15]: question for you, would  'ContentViewVersionExport' be a better name for that?   or just "Export', or something clearer that that class is export related (yeah i know it came from your previous pr)

File: app/services/katello/pulp3/content_view_version.rb [hunk 4]
```diff
@@ -15,10 +16,14 @@ def generate_exporter_id
       end
 
       def generate_exporter_path
-        export_path = "#{@content_view_version.content_view}/#{@content_view_version.version}".gsub(/\s/, '_')
+        export_path = "#{@content_view_version.content_view}/#{@content_view_version.version}/#{@destination_server}/#{increment_path}".gsub(/\s/, '_')
```
Comment (jlsherrill) [inline app/services/katello/pulp3/content_view_version.rb:19]: would a datestamp be more meaningful than an incrementing number?    2020-02-01-12-34?

File: app/controllers/katello/api/v2/content_view_versions_controller.rb (hunk 2 above)
Comment (parthaa) [reply app/controllers/katello/api/v2/content_view_versions_controller.rb:88]: not required

File: app/lib/actions/pulp3/content_view_version/create_exporter.rb (hunk 3 above)
Comment (parthaa) [reply app/lib/actions/pulp3/content_view_version/create_exporter.rb:15]: Well I plan to use it for import also... Not sure CVVE is a better name

Comment (jlsherrill) [reply app/lib/actions/pulp3/content_view_version/create_exporter.rb:15]: alright, keep it as is for now.  If you get to import though, think about if it really belongs in that service class, or if a 2nd one makes more sense.

Comment (parthaa) [reply app/lib/actions/pulp3/content_view_version/create_exporter.rb:15]: Yeh totally agree..

File: app/controllers/katello/api/v2/content_view_versions_controller.rb (hunk 1 above)
Comment (parthaa) [reply app/controllers/katello/api/v2/content_view_versions_controller.rb:94]: Add content view version also.

**Comment by parthaa** [comment]
```
$ curl -X GET -s -k -u admin:changeme "http://$(hostname -f):3000/katello/api/content_view_versions/export_histories?content_view_id=3"|jq
{
  "total": 1,
  "subtotal": 1,
  "page": 1,
  "per_page": 20,
  "error": null,
  "search": null,
  "sort": {
    "by": "id",
    "order": "asc"
  },
  "results": [
    {
      "destination_server": "life",
      "path": "/var/lib/pulp/exports/Default_Organization/wow/1.0/life/2020-08-26T21-20-46-00-00",
      "id": 4,
      "created_at": "2020-08-26 21:20:58 UTC",
      "updated_at": "2020-08-26 21:20:58 UTC",
      "content_view_version": "wow 1.0",
      "content_view_version_id": 3
    }
  ]
}
```

File: app/lib/actions/pulp3/content_view_version/destroy_exporter.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    module ContentViewVersion
+      class DestroyExporter < Pulp3::Abstract
+        input_format do
+          param :smart_proxy_id, Integer
+          param :exporter_data, Hash
+        end
+
+        def run
+          ::Katello::Pulp3::ContentViewVersion.new(smart_proxy: smart_proxy).destroy_exporter(input[:exporter_data][:pulp_href])
+        end
```
Comment (jlsherrill) [inline app/lib/actions/pulp3/content_view_version/destroy_exporter.rb:12]: is this an async action?

Comment (parthaa) [reply app/lib/actions/pulp3/content_view_version/destroy_exporter.rb:12]: not according to https://docs.pulpproject.org/pulpcore/restapi.html#operation/exporters_core_pulp_exports_delete

**Comment by parthaa** [comment]
[test katello]

**Comment by parthaa** [comment]
[test katello]

**Comment by parthaa** [comment]
[test katello]

**Comment by parthaa** [comment]
[test katello]

**Pull request merged**